File: Week-4/Day-1/Problem2.java

// Given an integer array nums, return the maximum result of nums[i] XOR nums[j], where 0 <= i <= j < n.

// Example 1:

// Input: nums = [3,10,5,25,2,8]
// Output: 28
// Explanation: The maximum result is 5 XOR 25 = 28.
// Example 2:

// Input: nums = [14,70,53,83,49,91,36,80,92,51,66,70]
// Output: 127
 

// Constraints:

// 1 <= nums.length <= 2 * 105
// 0 <= nums[i] <= 231 - 1



// https://leetcode.com/problems/maximum-xor-of-two-numbers-in-an-array/


class Solution {
    public int findMaximumXOR(int[] nums) {
        TrieNode root = new TrieNode();
        int maxXor = 0;

        for (int num : nums) {
            TrieNode node = root;
            TrieNode xorNode = root;
            int currentXor = 0;

            for (int i = 31; i >= 0; i--) {
                int bit = (num >> i) & 1;

                if (node.children[bit] == null) {
                    node.children[bit] = new TrieNode();
                }
                node = node.children[bit];

                int oppositeBit = 1 - bit;
                if (xorNode.children[oppositeBit] != null) {
                    currentXor |= (1 << i);
                    xorNode = xorNode.children[oppositeBit];
                } else {
                    xorNode = xorNode.children[bit];
                }
            }
            maxXor = Math.max(maxXor, currentXor);
        }

        return maxXor;
    }

    static class TrieNode {
        TrieNode[] children = new TrieNode[2];
    }
}

File: Week-4/Day-3/Problem2.py
# Given an m x n board of characters and a list of strings words, return all words on the board.

# Each word must be constructed from letters of sequentially adjacent cells, where adjacent cells are horizontally or vertically neighboring. The same letter cell may not be used more than once in a word.

 

# Example 1:


# Input: board = [["o","a","a","n"],["e","t","a","e"],["i","h","k","r"],["i","f","l","v"]], words = ["oath","pea","eat","rain"]
# Output: ["eat","oath"]
# Example 2:


# Input: board = [["a","b"],["c","d"]], words = ["abcb"]
# Output: []
 

# Constraints:

# m == board.length
# n == board[i].length
# 1 <= m, n <= 12
# board[i][j] is a lowercase English letter.
# 1 <= words.length <= 3 * 104
# 1 <= words[i].length <= 10
# words[i] consists of lowercase English letters.
# All the strings of words are unique.



# https://leetcode.com/problems/word-search-ii/



class TrieNode:
    def __init__(self):
        self.children = {}
        self.isEndOfWord = False

class Solution:
    def findWords(self, board: List[List[str]], words: List[str]) -> List[str]:
        root = TrieNode()
        
        def insert(word):
            node = root
            for char in word:
                if char not in node.children:
                    node.children[char] = TrieNode()
                node = node.children[char]
            node.isEndOfWord = True
        
        for word in words:
            insert(word)
        
        result = set()
        rows, cols = len(board), len(board[0])

        def dfs(i, j, node, path):
            if node.isEndOfWord:
                result.add(path)
                node.isEndOfWord = False  # To avoid duplicate results

            if i < 0 or i >= rows or j < 0 or j >= cols or board[i][j] not in node.children:
                return
            
            temp = board[i][j]
            board[i][j] = "#"
            nextNode = node.children[temp]

            dfs(i + 1, j, nextNode, path + temp)
            dfs(i - 1, j, nextNode, path + temp)
            dfs(i, j + 1, nextNode, path + temp)
            dfs(i, j - 1, nextNode, path + temp)

            board[i][j] = temp

        for i in range(rows):
            for j in range(cols):
                if board[i][j] in root.children:
                    dfs(i, j, root, "")

        return list(result)